examples/pytorch/rgcn/utils.py

```diff
@@ -334,3 +343,326 @@ def calc_mrr(embedding, w, train_triplets, valid_triplets, test_triplets, hits=[
+        entity_dict = {}
+        relation_dict = {}
+        node_types = []
+        graph_relations = set()
+        def handle_triples(triplets):
+            heads = []
+            head_types = []
+            rel_types = []
+            tails = []
+            tail_types = []
+            for triplet in triplets:
+                head_type, head_name = handle_node(triplet[0])
+                if head_type is None:
+                    continue
+                tail_type, tail_name = handle_node(triplet[2])
+                if tail_type is None:
+                    continue
+                # here the rel type is already (h, r, t)
+                rel_type = (head_type, triplet[1], tail_type)
+
+                rel_id = get_id(relation_dict, rel_type)
+                ed_len = len(entity_dict)
+                head_id = get_id(entity_dict, triplet[0])
+                head_type_id = get_id(entity_type_dict, head_type)
+                if len(entity_dict) > ed_len:
+                    node_types.append(head_type_id)
+                ed_len = len(entity_dict)
+                tail_id = get_id(entity_dict, triplet[2])
+                tail_type_id = get_id(entity_type_dict, tail_type)
+                if len(entity_dict) > ed_len:
+                    node_types.append(tail_type_id)
+
+                heads.append(head_id)
+                head_types.append(head_type_id)
+                rel_types.append(rel_id)
+                tails.append(tail_id)
+                tail_types.append(tail_type_id)
+                graph_relations.add((head_type_id, rel_id, tail_type_id))
+
+            heads = np.asarray(heads)
+            tails = np.asarray(tails)
+            head_types = np.asarray(head_types)
+            rel_types = np.asarray(rel_types)
+            tail_types = np.asarray(tail_types)
+            return heads, tails, head_types, rel_types, tail_types
+
+        train_head, train_tail, train_head_type, train_rel_type, train_tail_type = \
+            handle_triples(train_triplets)
+
+        valid_head, valid_tail, valid_head_type, valid_rel_type, valid_tail_type = \
+            handle_triples(valid_triplets)
+
+        test_head, test_tail, test_head_type, test_rel_type, test_tail_type = \
+            handle_triples(test_triplets)
+
+        self.train = (train_head, train_tail, train_head_type, train_rel_type, train_tail_type)
+        self.valid = (valid_head, valid_tail, valid_head_type, valid_rel_type, valid_tail_type)
+        self.test = (test_head, test_tail, test_head_type, test_rel_type, test_tail_type)
+        self.rel_map = relation_dict
+        self.entity_map = entity_dict
+        self.entity_type_dict = entity_type_dict
+        self.num_rels = len(list(graph_relations))
+        self.num_nodes = len(entity_dict)
+        self.node_types = np.asarray(node_types)
+
+def build_multi_ntype_heterograph_in_homogeneous_from_triplets(num_nodes, num_rels, node_types, edge_lists, reverse=True):
```

Comment:
Is this the same as `dgl.to_homo`?

Reply:
to_homo may not work for link prediction. There will be three graphs train_g, valid_g, test_g
Or I did not find a easy way to use to_homo.